src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs

```diff
@@ -694,6 +694,17 @@ private void SetGenParamCustomAttributeNoLock(CustAttr ca)
             m_ca ??= new List<TypeBuilder.CustAttr>();
             m_ca.Add(ca);
         }
+
+        internal void CreateGlobalModuleType()
```

Comment:
Who's using this?

Reply:
I implemented the changes outlined by @Jehoel in the [issue](https://github.com/dotnet/runtime/issues/68840).  

> > @lonix1 Under what conditions can `TypeBuilder.CreateType()` return null?
> 
> **TL;DR: There is no situation** when `TypeBuilder.CreateType()` will return `null` when `CreateType()` is being called by user-code. You can safely add a null-forgiving `!` to a `CreateType()` call-site.
> 
> The current source is at
> 
> https://github.com/dotnet/runtime/blob/1a296c06fff8750b4658f6f7d901347e006744d0/src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilder.cs#L1853
> 
>     * A quick browse of the current source  shows that `TypeBuilder.CreateType()` only returns `null` when the `TypeBuilder` is the builder for an assembly's hidden `<Module>` class.
>       
>       * The `<Module>` class is a special type that allows assemblies/modules to have eager-initialized globals.
>       * The C# language itself doesn't use `<Module>`, though other languages do (e.g. C++/CLI).
>       * Since C# 9.0, [you _can_ now opt-in to adding `<Module>` members](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/module-initializers) via the `[ModuleInitializer]` attribute.
>       * Of course, if you're using `Reflection.Emit` then you will have always been able to use `ModuleBuilder` going back to .NET 1.x in 2001.
> 
>     * `<Module>`-building `TypeBuilder` instances cannot be directly instantiated by user-code as it requires a specific `internal TypeBuilder()` constructor invocation - instead this ctor is called only when you use the `ModuleBuilder` type.
> 
>     * The only situation when `TypeBuilder.CreateType()` method _does_ indeed return `null` is inside `void ModuleBuilder.CreateGlobalFunctions()`, which basically ensures that a `class <Module>` _will_ be defined in the generated dynamic-assembly.
> 
> 
> So based on that, I think the design of `TypeBuilder`'s API should be changed in a minimally-breaking way:
> 
>     * Change the current `public Type? TypeBuilder.CreateType()` method into `private Type? TypeBuilder.CreateTypeImpl()`.
> 
>     * Add back `public Type TypeBuilder.CreateType()` (without the nullable return type) which just calls `CreateTypeImpl()` directly, maybe adding a debug assertion that `CreateTypeImpl` _did not_ return `null`?).
> 
>     * Add a new method to `TypeBuilder`: `internal void CreateGlobalModuleType()` which first checks that `this.m_isHiddenGlobalType == true` (and if not throws, I suppose) and then calls `CreateTypeImpl()` with a debug assertion that it _did_ return `null`.
> 
> 
> Alternatively, refactor the `CreateType()` method more thoroughly?

Reply:
> I implemented the changes outlined by @Jehoel in the https://github.com/dotnet/runtime/issues/68840.

But presumably the intention is that there's some existing caller of CreateType that would need to be changed to use CreateGlobalModuleType?

Reply:
Yes, but not only that, the `CreateGlobalModuleType()` method unconditionally throws so the call never gonna succeed.

Reply:
Ah, I've fixed the `throw` in `CreateGlobalModuleType()` and replaced the method call for `_globalTypeBuilder`.